File: python/0557.py
n = input()
arr = map(int, raw_input().split())
answer = arr[n-1]

dp = [[0]*21 for i in range(n)]
dp[0][arr[0]] = 1

for i in range(1, n-1):
  for j in range(21):
    if dp[i-1][j] == 0: continue
    vp = j + arr[i]
    vm = j - arr[i]

    if vp <= 20:
      dp[i][vp] += dp[i-1][j]
    if vm >= 0:
      dp[i][vm] += dp[i-1][j]

print dp[n-2][answer]

File: ruby/0557.rb
# 動的計画法(DP)

loop do
  n = gets.to_i
  break if n == 0
  arr = gets.split(' ').map(&:to_i)

  dp = Array.new(n) { Array.new(21, 0) }

# スタート地点
  dp[0][arr[0]] = 1

  1.upto(n-2) do |i|
    0.upto(20) do |j|
      next if dp[i-1][j] == 0

      plus = j + arr[i]
      minus = j - arr[i]

      dp[i][plus] += dp[i-1][j] if plus <= 20
      dp[i][minus] += dp[i-1][j] if minus >= 0
    end
  end

  puts dp[n-2][arr[n-1]]
end
